README
StatPatternRecognition README
-----------------------------

Please see the documentation files in the doc directory (specifically
index.html/index.txt) for information about obtaining, contributing to, and
running SPR.


HISTORY
$Id: HISTORY,v 1.56 2008-05-09 21:25:25 narsky Exp $

05.09.2008 narsky V08-02-00
	Minor changes. Added an extra handle to SprGEP. Modified the
	random hill climbing algorithm for the matrix generation to
	invert rows as well, not only columns. Fixed the author list.

05.08.2008 narsky V08-01-00
	Piti Ongmongkolkul implemented a multi-class learner matrix
	generator capable of optimizing row or column Hamming
	distance. For detail, see README section 2a.4.

	Rules of ownership of weak classifiers by super-engines such
	as AdaBoost or Bagger have been changed to reduce memory
	consumption. As a consequence all makeTrained() methods are
	now non-const. The user can revert to the old design by
	setting SprEnforceLowMemory flag to false. For detail see
	section 17 in README.

	Cleaned up SprGEP and SprChromosome implementations. The GEP
	algorithm is now ready for mass consumption.

	Fixed bugs in SprFisher and SprLogitR that prevented these
	methods from being used by the "add n remove r" algorithm.

	Fixed a bug in SprPreFilter that prevented variable
	transformations from being applied correctly in some cases.

	Cleaned up README.

	Various fixes elsewhere.

04.02.2008 narsky V08-00-00
	Implemented a new multiclass learning algorithm, "binary
	encoder". See section 2a.2 in README. The new method is only
	accessible through SprMultiClassApp but not (yet) from the
	Root interface.

	Changed format used for storing trained multiclass learner
	configurations. The user will not be able to use multiclass
	learner configuration files produced with earlier versions.

	Added several options to SprMultiClassApp: adjustment of
	weights for classifiers and exclusion of neutral classes
	(indicated by zero's in the class-classifier matrix) from the
	loss computation for each class. See section 2a.1 in README.

	The multiclass learner can now compute missing values for each
	input class separately and assign class labels to new data by
	replacing missing values with medians for each class in
	turn. See section 2a.3 in README.

	The framework for treatment of missing values has been
	changed. Instead of letting every method replace missing
	values on fly with '-m' command-line option, the package applies
	a data transformation in the same way as PCA or normalization
	of variables to the same input range. First, you need to
	compute medians or average of 1D marginal densities through
	SprTransformationApp and then apply the computed
	transformation using '-Q' option. See section 15 in README.

02.29.2008 narsky V07-15-00
	Changed the splitting algorithm for cross-validation to allow
	for any number of CV pieces. The algorithm is now capable of
	dividing the dataset into subsets efficiently for any number
	of subsets including extreme cases such as the "leave one out" 
	CV method.

	Introduced new interfaces for SprFomCalculator. One does not
	have to supply the full list of trained classifiers for CV -
	instead one only needs to supply a vector of responses 
	computed for all the trained classifiers. SprCrossValidator
	now processes the trained classifier for each subset separately
	and deletes it immediately after responses are filled to avoid
	excessive memory consumption.

	Added classBlind flag to SprAbsFilter::replaceMissing() method.
	If set to false, it will compute missing values as medians
	of distributions for a specific class. This flag is set to true
	in all executables by default.

	Reduced verbosity of some classifiers to make output from 
	SprAddNRemoveRApp more pleasant.

	SprIOTestApp now displays full info about all classes found
	in the input data.

02.25.2008 narsky V07-14-00
	Implemented a better splitting algorithm for dividing data into
	pieces for cross-validation. The pieces are now evenly divided
	across different classes (with an arbitrary number of classes
	allowed). The leftovers are added to the last piece.

	Implemented cross-validation for the multiclass learner. Invoked
	by -x option of SprMultiClassApp.

	Implemented "integrate" option for cross-validation. If the
	"integrate" flag is set to true, FOM is integrated over all
 	points in the supplied data. If "integrate" is false, a
 	separate FOM is computed for each validation piece and then an
 	average of these FOMs is estimated and an error is derived
 	from the spread of the FOM values around the average. If
 	"integrate" is true, no error is estimated.

	Fixed a bug in SprTrainedMultiClassLearner::classes() - due to
	some strange reason it returned an ordered list of classes
	which, of course, screwed up the classification table returned
	by SprMultiClassApp if the input classes were not supplied
	in the same order.

02.22.2008 narsky V07-13-00
	Fixed a bug in computation of SprLoss::correct_id (the default
	choice for SprAddNRemoveRApp).

	Fixed a bug in cross-validation settings for SprDecisionTreeApp:
	the classes for the cross-validator were always set to 0 and 1
	by mistake. If the user attempted to run CV with other classes,
	this would give wrong results. CV in SprBaggerDecisionTreeApp
	and SprAdaBoostDecisionTreeApp was free of this bug.

	Added a new option, -C, to the 3 executables: SprDecisionTreeApp, 
	SprBaggerDecisionTreeApp and SprAdaBoostDecisionTreeApp. Now -c
	option chooses FOM to be optimized by the decision tree, and -C
	chooses FOM to be monitored for cross-validation or validation
	data. The default settings are: Gini index for optimization and
	accuracy (correctly classified fraction) for monitoring.

	Introduced SprCrossValidatorApp, an executable capable of
	running cross-validation with any classifier included in the
	package. The input configuration file follows the same syntax
	as booster.config.

	useStandard() and useNormalized() methods of trained classifiers
	now change the default cut on the classifier output when
	appropriate: 0.0 for useStandard() and 0.5 for useNormalized().
	The cut from the trainable classifier is now forced onto the 

root/spr_plot.C
// $Id: spr_plot.C,v 1.7 2007-12-06 00:04:10 narsky Exp $

void plotEffCurve(const char* canvas,
		  int npts, const double* signalEff, const char* classifier,
		  const double* bgrndEff, const double* bgrndErr,
		  int icolor) 
{
  TCanvas *c 
    = new TCanvas(canvas,"SPR BgrndEff vs SignalEff",200,10,600,400);
  TGraph *gr = new TGraphErrors(npts,signalEff,bgrndEff,0,bgrndErr);
  TLegend *leg = new TLegend(0.1,0.85,0.5,1.,"SPR Classifiers","NDC");
  gr->SetMarkerStyle(21);
  gr->SetLineColor(icolor+1);
  gr->SetLineWidth(3);
  gr->SetMarkerColor(icolor+1);
  gr->Draw("ALP");
  leg->AddEntry(gr,classifier,"P");
  leg->Draw();
}


void plotFOMCurve(const char* canvas, const char* title,
		  int npts, const double* signalEff, const char* classifier,
		  const double* fom, int icolor) 
{
  TCanvas *c 
    = new TCanvas(canvas,"SPR FOM vs SignalEff",200,10,600,400);
  TGraph *gr = new TGraphErrors(npts,signalEff,fom,0,0);
  TLegend *leg = new TLegend(0.1,0.85,0.5,1.,title,"NDC");
  gr->SetMarkerStyle(21);
  gr->SetLineColor(icolor+1);
  gr->SetLineWidth(3);
  gr->SetMarkerColor(icolor+1);
  gr->Draw("ALP");
  leg->AddEntry(gr,classifier,"P");
  leg->Draw();
}


void plotEffCurveMulti(const char* canvas,
		       int ngraph, int npts, const double* signalEff, 
		       const char classifiers[][200],
		       const double* bgrndEff, const double* bgrndErr,
		       double ymax, bool setMax=false)
{
  TCanvas *c
    = new TCanvas(canvas,"SPR BgrndEff vs SignalEff",200,10,600,400);
  TMultiGraph *mg = new TMultiGraph();
  if( setMax ) mg->SetMaximum(ymax);
  TLegend *leg = new TLegend(0.1,0.85,0.5,1.,
			     "SPR BackgroundEff vs SignalEff","NDC");
  for( int i=0;i<ngraph;i++ ) {
    TGraph *gr = new TGraphErrors(npts,signalEff,
				  bgrndEff+(i*npts),0,bgrndErr+(i*npts));
    gr->SetMarkerStyle(21);
    gr->SetLineColor(i+1);
    gr->SetLineWidth(3);
    gr->SetMarkerColor(i+1);
    mg->Add(gr);
    leg->AddEntry(gr,classifiers[i],"P");
  }
  mg->Draw("ALP");
  leg->Draw();
}


void plotCorrelation(const char* canvas, const char* title,
		     const unsigned nVars, const char vars[][200], 
		     const double* corr)
{
  TCanvas *c
    = new TCanvas(canvas,"SPR Variable Correlations",200,10,600,400);
  gStyle->SetPalette(1);
  gStyle->SetOptTitle(1);
  gStyle->SetOptStat(0);
  TH2D* h2 = new TH2D("corr",title,nVars,0,nVars,nVars,0,nVars);
  for( int i=0;i<nVars;i++ ) {
    for( int j=0;j<nVars;j++ ) {
      h2->Fill(vars[i],vars[j],corr[i*nVars+j]);
    }
  }
  h2->SetLabelSize(0.06,"X");
  h2->SetLabelSize(0.06,"Y");
  h2->LabelsDeflate("X");
  h2->LabelsDeflate("Y");
  h2->LabelsOption("v");
  h2->Draw("Z COL");
}


void plotImportance(const char* canvas, const char* title,
		    const unsigned nVars,
		    const char vars[][200], 
		    const double* importance, const double* error,
		    bool automaticRange=false)
{
  TCanvas *c 
    = new TCanvas(canvas,"SPR Variable Importance",200,10,600,400);
  gStyle->SetPalette(1);
  TH1D* h1 = new TH1D("importance",title,nVars,0,nVars);
  double ymin(0), ymax(0);
  for( int i=0;i<nVars;i++ ) {
    h1->Fill(vars[i],importance[i]);
    if( automaticRange ) {
      ymin = ( ymin>importance[i] ? importance[i] : ymin );
      ymax = ( ymax<importance[i] ? importance[i] : ymax );
    }
  }
  if( automaticRange ) {
    ymin = ( ymin<0 ? ymin*1.2 : ymin*0.8 );
    ymax *= 1.2;
  }
  else {
    ymin = 0.;
    ymax = 1.;
  }
  if( error == 0 ) {
    for( int i=0;i<nVars;i++ )
      h1->SetBinError(i+1,0.);
  }
  else {
    for( int i=0;i<nVars;i++ )
      h1->SetBinError(i+1,error[i]);
  }
  h1->LabelsDeflate("X");
  h1->LabelsOption("v");
  h1->SetLabelSize(0.06,"X");
  h1->SetLineColor(4);
  h1->SetMarkerStyle(21);
  h1->SetMarkerColor(4);
  h1->SetLineWidth(2);
  TAxis* yaxis = h1->GetYaxis();
  yaxis->SetRangeUser(ymin,ymax);
  if( error == 0 ) {
    h1->SetLineColor(4);
    h1->SetFillColor(4);
    h1->SetBarWidth(0.8);
    h1->SetBarOffset(0.1);
    h1->Draw("B");
  }
  else
    h1->Draw("E0P");
  l = new TLine(0,0,nVars,0); l->Draw();
}


// Modes: linear or log
plotHistogram(const char* canvas, const char* mode, const char* title,
	      double xlo, double xhi, int nbin,
	      double* sig, double* sigerr, double* bgr, double* bgrerr)
{
  TCanvas *c 
    = new TCanvas(canvas,"SPR Classifier Output",200,10,600,400);
  gStyle->SetPalette(1);
  TLegend *leg = new TLegend(0.1,0.85,0.5,1.,"Classifier Output","NDC");
  double dx = (xhi-xlo) / nbin;
  TH1D* hs = new TH1D("signal",    title,nbin,xlo,xhi);
  TH1D* hb = new TH1D("background",title,nbin,xlo,xhi);
  leg->AddEntry(hs,"Signal","L");
  leg->AddEntry(hb,"Background","L");
  for( int i=0;i<nbin;i++ ) {
    hs->SetBinContent(i+1,sig[i]);
    hs->SetBinError(i+1,sigerr[i]);
    hb->SetBinContent(i+1,bgr[i]);
    hb->SetBinError(i+1,bgrerr[i]);
  }
  TPad* pad = new TPad("pad","pad",0,0,1,1);
  if( strcmp(mode,"log") == 0 ) pad->SetLogy(1);
  pad->Draw();
  pad->cd();
  hs->SetLineColor(2);
  hs->SetLineWidth(3);
  hb->SetLineColor(4);
  hb->SetLineWidth(3);
  hb->Draw();
  hs->Draw("same");
  leg->Draw();
}


plotClasses(const char* canvas, const char* title,
	    int nClasses, const char classes[][200], 
	    const int* events, const double* weights)
{
  char evtitle[200], wttitle[200];
  strcpy(evtitle,title);
  strcpy(wttitle,title);
  strcat(evtitle,": Events");
  strcat(wttitle,": Weights");

  TCanvas *c 
    = new TCanvas(canvas,"SPR Input Classes",200,10,600,400);
  gStyle->SetPalette(1); 

  int maxEv = TMath::MaxElement(nClasses,events);
  double maxWt = TMath::MaxElement(nClasses,weights);

  TPad* pad1 = new TPad("events", evtitle,0,0,1.,0.5);
  TPad* pad2 = new TPad("weights",wttitle,0,0.5,1.,1.);
  pad1->Draw();
  pad2->Draw();

  // events
  pad1->cd();
  TH1I* hev = new TH1I("events",evtitle,nClasses,0,nClasses);
  for( int i=0;i<nClasses;i++ )
    hev->Fill(classes[i],events[i]);
  hev->LabelsDeflate("X");
  hev->SetLabelSize(0.06,"X");
  hev->SetLabelSize(0.1,"X");
  hev->SetLabelSize(0.1,"Y");
  hev->SetLineColor(4);
  hev->SetFillColor(4);
  hev->SetBarWidth(0.8);
  hev->SetBarOffset(0.1);
  TAxis* yaxis1 = hev->GetYaxis();
  yaxis1->SetRangeUser(0.,1.1*maxEv);
  hev->Draw("B");

  // weights
  pad2->cd();
  TH1D* hwt = new TH1D("weights",wttitle,nClasses,0,nClasses);
  for( int i=0;i<nClasses;i++ )
    hwt->Fill(classes[i],weights[i]);
  hwt->LabelsDeflate("X");
  hwt->SetLabelSize(0.06,"X");
  hwt->SetLabelSize(0.1,"X");
  hwt->SetLabelSize(0.1,"Y");
  hwt->SetLineColor(3);
  hwt->SetFillColor(3);
  hwt->SetBarWidth(0.8);
  hwt->SetBarOffset(0.1);
  TAxis* yaxis2 = hwt->GetYaxis();
  yaxis2->SetRangeUser(0.,1.1*maxWt);
  hwt->Draw("B");
}


void plotMultiClassTable(const char* canvas, const char* title,
			 const int nClass, const char classes[][200], 
			 const double* mcTable)
{
  TCanvas *c = 
    new TCanvas(canvas,"SPR Assigned Class Label (Y) vs True Class Label (X)",
		200,10,600,400);
  gStyle->SetPalette(1);
  gStyle->SetOptTitle(1);
  gStyle->SetOptStat(0);
  TH2D* h2 = new TH2D("MultiClass",title,nClass,0,nClass,nClass,0,nClass);
  for( int i=0;i<nClass;i++ ) {
    for( int j=0;j<nClass;j++ ) {
      h2->Fill(classes[i],classes[j],mcTable[i*nClass+j]);
    }
  }
  h2->LabelsDeflate("X");
  h2->LabelsDeflate("Y");
  h2->SetLabelSize(0.06,"X");
  h2->SetLabelSize(0.06,"Y");
  h2->LabelsOption("v");
  h2->Draw("Z COL");
}
